### ISSUE TITLE
Missing main.cpp and Source Files

### ISSUE THREAD

**Issue by SteveBenner** [Mr-Twave/Open_Spaceward_Ho-#1]
Hi, I attempted to build the project using `cmake`, but encountered an error related to missing `main.cpp` and source files referenced in `CMakeLists.txt`. Could you clarify if these files are included in the repository or if additional steps are needed to build the project?

How close to being able to build the software so as to be playable exactly is this project?

Thanks!

I may be interesting in helping further develop this project.

**Comment by Mr-Twave** [comment]
Hi,
This project has been on pause for some bit because I've had some time management issues to take care of. I'm close to the other end of that, however.
You see the file structure of this through the README- I actually have several versions of this repo locally. I've got about 2x more code than I've been sharing, with lots of code missing on the GUI and assets front. I'd like to make this game *look* much like the original Spaceward Ho! But I can settle for something similar as well. I plan to leverage some ML techniques on the AI s.t. the player is actually difficult to play against, preferably using a very light transformer to do so. (~220 M parameters- easily runnable and trainable on smaller computers).
Perhaps later this week I'll upload all of the codebase that I have so far- it's much more vast and disorganized than what you see currently on this GitHub. Make no mistake, this project is nowhere near it's goal- but it possesses the 'necessity' in logic and in game vectors, theoretically making this game very lightweight and easy to bugfix/learnable for AI. I've been creating tools for myself s.t. I can organize my code easier, so when those roll out (which should now be very soon, in under a week or two, actually!) I can share something... MUCH easier to work on.

**Comment by melissa-anholm** [comment]
Hello!  I'd like to add another vote for "I'd probably like to contribute to this".

**Comment by Mr-Twave** [comment]
I'll put an exact timeline on what to expect:

December 21st - I'll release my extensive ~400 re-organized cpp function set.
December 22nd - I'll let in contributors in on the project.
December 30th - I'll create a small discord server (if we are interested in making one?) for following through with the project's creation.

**Comment by melissa-anholm** [comment]
Welp.  I'm excited.  

Follow-up question:  what AI/ML packages/techniques were you thinking to use for this?  Of course the AI training could just be baked in, but it'd be super cool to have the option for an AI that learns your play style.  

I'd have to think a bit harder about how I'd want to build and train the model (if it were left up to me obviously, which it isn't), but it seems totally do-able, and I'm sure you've already figured a lot of this out.

**Comment by Mr-Twave** [comment]
I could say it's really simple, but really you just need to understand anything the human player would 'see' should be something visible to the AI player as well.

I'm only speculating what the AI is going to look like, but it should be relatively straightforward to train. We do not need a very large model, only a bunch of really small ones- thus should be really easy to train each. The complexity of this game is high, but we can use existent architectures to make our problems *much much* easier.

As such, you create something like this (of course, this vector would be most simply utilize an encoder-only transformer or similar);
- A vector state can be input
- The transformer can 'pay attention' to particular parts of the vector state while it makes decisions of which are fixed based upon the game
- The transformer gets input information about previous games (of which, is something I will not yet disclose until December 30th.)

The encoding of the transformer's inputs can be the exact expectation of what a player sees. Thus, the logic of the game and the logic of the transformer can rely on the same base cases.


Roughly put, the game and the AI transformers can both rely on through training;
- State-to-Vector representing Playerstate
- State-to-Vector representing playerassets
- State-to-Vector representing Lastseen Opponentasset (player-number associated)
- State-to-Vector representing unseenassets (information about the rest of the galaxy)
- State-to-Vector representing NumberOfPlayers
- Boolean representing Playerisdead (to note the removal of a player)
- Integer noting playerturn (of which gets noised up in training data during training to avoid overfitting)

Notably the training data would be somewhat computationally expensive to produce. So simply, we can try smaller games before we make larger games. (e.g. 3-4 planet galaxy simulation rollouts) as its own base-case (noising up the training data s.t. the transformer will believe it does not matter that there are other planets), then gradually increase the complexity during the training process s.t. the transformer will take advantage of the larger space of the game without having to 'stretch' the model via overfitting.

The transformer can be 'seeded' with a partition of a larger transformer, or an already existent 1-2M encoder transformer of which we can reduce its output tokenization space to 'W/L' predictions as well as numbers, as well as a vector P that suggests the probability states of the transformer's outputs. Any game actions can be repurposed to span a subsection of the tokenization space.

This process is probably best done for really small transformers (1-2 million parameters) then we increase the amount of nullified blocks within the transformers s.t. we are increasing complexity over time so that we can train this AI with minimal computational resources. Ideally, we'd actually use a LLM encoder s.t. we take an open weights large language model and use it to seed our encoder.

We also want the AI to produce a series of expectation states. Ideally, to prevent overfitting, we can noise up the training process some.

We can use a Google Collab project to simulate the environment needed to simulate the games. 

Resources (we do not need to create a million-paremter LLM from scratch- though we can create a transformer using a similar process, perhaps a transformer);

https://github.com/FareedKhan-dev/create-million-parameter-llm-from-scratch/blob/main/code.ipynb

**Comment by SteveBenner** [comment]
How is the progress? I'm interested in a Discord server if you are going to set one up. That way we can facilitate better collaboration for sure.

**Comment by OrbitalBliss** [comment]
Yes, I'm interested in the project as well.
Maybe I can help with asset generation using AI or something.
If there's a discord or such, lemme know.

Good luck

**Comment by SteveBenner** [comment]
What's the status on this guy?
